src/compiler/cse.rs

```diff
@@ -0,0 +1,452 @@
+use rand_chacha::{rand_core::SeedableRng, ChaCha20Rng};
+
+use std::{
+    collections::{HashMap, HashSet},
+    hash::Hash,
+    marker::PhantomData,
+};
+
+use crate::{
+    field::Field,
+    poly::{
+        self,
+        cse::{create_common_ses_signal, replace_expr},
+        Expr, HashResult, VarAssignments,
+    },
+    sbpir::{
+        query::Queriable, sbpir_machine::SBPIRMachine, ForwardSignal, InternalSignal, StepType,
+    },
+    wit_gen::NullTraceGenerator,
+};
+
+#[derive(Clone, Debug)]
+pub struct CseConfig {
+    max_iterations: usize,
+}
+
+impl Default for CseConfig {
+    fn default() -> Self {
+        Self {
+            max_iterations: 100,
+        }
+    }
+}
+
+#[allow(dead_code)]
+pub fn config(max_iterations: Option<usize>) -> CseConfig {
+    CseConfig {
+        max_iterations: max_iterations.unwrap_or(100),
+    }
+}
+
+pub trait Scoring<F: Field + Hash> {
+    fn score(&self, expr: &Expr<F, Queriable<F>, HashResult>, info: &SubexprInfo) -> usize;
+}
+
+pub struct Scorer {
+    min_degree: usize,
+    min_occurrences: usize,
+}
+
+impl Default for Scorer {
+    fn default() -> Self {
+        Self {
+            min_degree: 2,
+            min_occurrences: 2,
+        }
+    }
+}
+
+impl<F: Field + Hash> Scoring<F> for Scorer {
+    fn score(&self, _expr: &Expr<F, Queriable<F>, HashResult>, info: &SubexprInfo) -> usize {
+        if info.degree < self.min_degree || info.count < self.min_occurrences {
+            0
+        } else {
+            info.count * info.degree
+        }
+    }
+}
+
+/// Common Subexpression Elimination (CSE) optimization.
+/// This optimization replaces common subexpressions with new internal signals for the step type.
+/// This is done by each time finding the optimal subexpression to replace and creating a new signal
+/// for it and replacing it in all constraints.
+/// The process is repeated until no more common subexpressions are found.
+/// Equivalent expressions are found by hashing the expressions with random assignments to the
+/// queriables. Using the Schwartz-Zippel lemma, we can determine if two expressions are equivalent
+/// with high probability.
+#[allow(dead_code)]
+pub(super) fn cse<F: Field + Hash, S: Scoring<F>>(
+    mut circuit: SBPIRMachine<F, NullTraceGenerator>,
+    config: CseConfig,
+    scorer: &S,
+) -> SBPIRMachine<F, NullTraceGenerator> {
+    for (_, step_type) in circuit.step_types.iter_mut() {
+        cse_for_step(step_type, &circuit.forward_signals, &config, scorer)
+    }
+    circuit
+}
+
+fn cse_for_step<F: Field + Hash, S: Scoring<F>>(
+    step_type: &mut StepType<F, ()>,
+    forward_signals: &[ForwardSignal],
+    config: &CseConfig,
+    scorer: &S,
+) {
+    let mut signal_factory = SignalFactory::default();
```

Comment:
@rutefig in general this function is too long. You should see how to split it in many small functions, and make this function read more like a script of all the different steps that it take